### PR TITLE
Correct clean up actions in e2e tests

### LINKED ISSUE #55505: Clean up actions are not used correctly in e2e tests
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

### Problem
In some e2e tests, `framework.AddCleanupAction()` is used:
https://github.com/kubernetes/kubernetes/blob/1a4a019e5f84637f125589288edca46a974c1486/test/e2e/apimachinery/webhook.go#L50-L52

These cleanup actions are expected to be called when tests are terminated, by calling `framework.RunCleanupActions()` in test/e2e/e2e.go. The problem is that, in test/e2e/e2e.go, duplicated add/remove/run cleanup actions codes are implemented, and the `RunCleanupActions()` in test/e2e is wrongly used:
https://github.com/kubernetes/kubernetes/blob/1a4a019e5f84637f125589288edca46a974c1486/test/e2e/e2e.go#L294-L301

This caused the aformentioned clean up actions added in e2e tests not being called.

### Solution
In test/e2e/e2e.go, we should remove the duplicate "cleanup action" code, and use the clean up code in test/e2e/framework instead.

However, e2e tests won't pass after this is fixed. It's because some e2e tests added clean up functions (functions added in`framework.AddCleanupAction()`) that don't work. For example, in test/e2e/apimachinery/webhook.go, `cleanWebhookTest(f)` uses `f *framework.Framework` as input, whereas `f` already becomes `nil` by the time clean up actions are called. This can be solved by storing required variables in `BeforeEach`:

https://github.com/kubernetes/kubernetes/blob/1a4a019e5f84637f125589288edca46a974c1486/test/e2e/apps/deployment.go#L67-L70

@kubernetes/sig-testing-bugs

### PR DESCRIPTION
**Pull request by miaoyq**
**What this PR does / why we need it**:
Remove the duplicate "cleanup action" code in `test/e2e/e2e.go`, and use the clean up code in test/e2e/framework instead.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #55505 

**Special notes for your reviewer**:

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by guangxuli** [comment]
@miaoyq  please sweep other e2e test case calling  `framework.AddCleanupAction.` :)

**Comment by miaoyq** [comment]
@guangxuli  Thanks for your reminder. :-)

File: test/e2e/e2e.go [hunk 1]
```diff
@@ -249,55 +248,12 @@ var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
...
-	}
-}
-
 // Similar to SynchornizedBeforeSuite, we want to run some operations only once (such as collecting cluster logs).
 // Here, the order of functions is reversed; first, the function which runs everywhere,
 // and then the function that only runs on the first Ginkgo node.
 var _ = ginkgo.SynchronizedAfterSuite(func() {
 	// Run on all Ginkgo nodes
 	framework.Logf("Running AfterSuite actions on all node")
-	RunCleanupActions()
```
Comment (janetkuo) [inline test/e2e/e2e.go]: call `framework.RunCleanupActions()` here

Comment (miaoyq) [reply test/e2e/e2e.go]: Thanks @janetkuo, Done.

**Comment by janetkuo** [comment]
Most failures are caused by merge conflicts. Should be fixed by rebase.

**Comment by miaoyq** [comment]
@janetkuo Rebased. :)

**Comment by janetkuo** [comment]
It needs another rebase now. Applying a tag to bump it in the submit queue to (hopefully) prevent more rebases.

File: test/e2e/apimachinery/aggregator.go [hunk 2]
```diff
@@ -47,9 +49,20 @@ import (
 var serverAggregatorVersion = utilversion.MustParseSemantic("v1.7.0")
 
 var _ = SIGDescribe("Aggregator", func() {
+	var ns string
+	var c clientset.Interface
+	var aggrclient *aggregatorclient.Clientset
 	f := framework.NewDefaultFramework("aggregator")
 	framework.AddCleanupAction(func() {
-		cleanTest(f)
+		if c != nil && aggrclient != nil {
```
Comment (janetkuo) [inline test/e2e/apimachinery/aggregator.go]: We can remove these nil checks. They shouldn't be `nil` if cleanup is called as expected

File: test/e2e/apimachinery/webhook.go [hunk 3]
```diff
@@ -66,9 +66,18 @@ const (
 var serverWebhookVersion = utilversion.MustParseSemantic("v1.8.0")
 
 var _ = SIGDescribe("AdmissionWebhook", func() {
+	var ns string
+	var c clientset.Interface
 	f := framework.NewDefaultFramework("webhook")
 	framework.AddCleanupAction(func() {
-		cleanWebhookTest(f)
+		if c != nil {
```
Comment (janetkuo) [inline test/e2e/apimachinery/webhook.go]: Same comment here about the nil check, and everywhere else

File: test/e2e/apimachinery/aggregator.go (hunk 2 above)
Comment (miaoyq) [reply test/e2e/apimachinery/aggregator.go]: @janetkuo  We may need the nil checks.
`f.ClientSet` isn't created by framework.NewDefaultFramework("aggregator") directly but created by `f.BeforeEach` , see:
https://github.com/kubernetes/kubernetes/blob/02b3928a921b2e0930bf67b48f71c9dcd1086b4f/test/e2e/framework/framework.go#L126-L138

if we run test case like:
```
go run hack/e2e.go -v -test --test_args='--ginkgo.focus="xxx"'
```
some test cases may be skipped, `f.BeforeEach` will not be called, but `framework.AddCleanupAction(func())` will be called. In this situation, the value of `f.ClientSet` will be `nil`.

**Comment by miaoyq** [comment]
>It needs another rebase now. Applying a tag to bump it in the submit queue to (hopefully) prevent more rebases.

Thanks @janetkuo , will rebase. :-)

**Comment by miaoyq** [comment]
@janetkuo  Rebased.

File: test/e2e/apimachinery/aggregator.go (hunk 2 above)
Comment (janetkuo) [reply test/e2e/apimachinery/aggregator.go]: Ah, you're right. However it seems wrong to have skipped test add cleanup actions. We should fix that (not here, in a follow up PR). 

`f.Namespace` should be set unless it's skipped. Just checking namespace should be enough (`if len(ns) > 0 {...}`).

```go
type Framework struct {
	Namespace                *v1.Namespace   // Every test has at least one namespace unless creation is skipped
}
```

Would you also add a comment here (and everywhere else) to explain this? Something like:
```go 
// Cleanup actions will be called even when the tests are skipped, leaving namespace unset.
```

**Comment by janetkuo** [comment]
Left a comment: https://github.com/kubernetes/kubernetes/pull/55525#discussion_r151284967

**Comment by miaoyq** [comment]
Thanks @janetkuo, will address the comment. :)

**Comment by miaoyq** [comment]
@janetkuo Addressed.

File: test/e2e/storage/vsphere_volume_placement.go [hunk 4]
```diff
@@ -77,11 +77,13 @@ var _ = SIGDescribe("Volume Placement", func() {
 		2. Delete VMDK volume
 	*/
 	framework.AddCleanupAction(func() {
-		if len(node1KeyValueLabel) > 0 {
-			framework.RemoveLabelOffNode(c, node1Name, "vsphere_e2e_label")
-		}
-		if len(node2KeyValueLabel) > 0 {
-			framework.RemoveLabelOffNode(c, node2Name, "vsphere_e2e_label")
+		if c == nil {
```
Comment (janetkuo) [inline test/e2e/storage/vsphere_volume_placement.go]: This should be `c != nil`.

Suggest changing this one to `len(ns)` and add the comments too

Comment (miaoyq) [reply test/e2e/storage/vsphere_volume_placement.go]: I'm sorry for missing this, have updated.

File: test/e2e/storage/vsphere_volume_diskformat.go [hunk 5]
```diff
@@ -73,15 +73,16 @@ var _ = SIGDescribe("Volume Disk Format [Feature:vsphere]", func() {
 			framework.Failf("Unable to find ready and schedulable Node")
 		}
 		if !isNodeLabeled {
-			nodeLabelValue := "vsphere_e2e_" + string(uuid.NewUUID())
+			nodeLabelValue = "vsphere_e2e_" + string(uuid.NewUUID())
```
Comment (janetkuo) [inline test/e2e/storage/vsphere_volume_diskformat.go]: good catch

**Review by janetkuo — APPROVED** [review]

**Comment by caesarxuchao** [comment]
Thanks @miaoyq. @cheftako miaoyq was a step ahead of you.

File: test/e2e/apimachinery/webhook.go [hunk 6]
```diff
@@ -563,11 +571,8 @@ func updateConfigMap(c clientset.Interface, ns, name string, update updateConfig
 	return cm, pollErr
 }
 
-func cleanWebhookTest(f *framework.Framework) {
-	client := f.ClientSet
+func cleanWebhookTest(client clientset.Interface, namespaceName string) {
 	_ = client.AdmissionregistrationV1alpha1().ValidatingWebhookConfigurations().Delete(webhookConfigName, nil)
-	_ = client.AdmissionregistrationV1alpha1().ValidatingWebhookConfigurations().Delete(crdWebhookConfigName, nil)
```
Comment (caesarxuchao) [inline test/e2e/apimachinery/webhook.go]: Why is this line removed?

File: test/e2e/apimachinery/webhook.go [hunk 7]
```diff
@@ -71,12 +71,20 @@ var serverWebhookVersion = utilversion.MustParseSemantic("v1.8.0")
 
 var _ = SIGDescribe("AdmissionWebhook", func() {
 	var context *certContext
+	var ns string
+	var c clientset.Interface
 	f := framework.NewDefaultFramework("webhook")
 	framework.AddCleanupAction(func() {
```
Comment (caesarxuchao) [inline test/e2e/apimachinery/webhook.go]: Oh, this fix is still not complete. The CleanupAction is called after the entire suite is done, but we need to call these cleanups after each `It` block. @cheftako has some code in his PR that's doing the right thing I think.

File: test/e2e/apimachinery/webhook.go [hunk 8]
```diff
@@ -71,12 +71,20 @@ var serverWebhookVersion = utilversion.MustParseSemantic("v1.8.0")
...
-		cleanWebhookTest(f)
+		// Cleanup actions will be called even when the tests are skipped and leaves namespace unset.
```
Comment (caesarxuchao) [inline test/e2e/apimachinery/webhook.go]: I thought the namespace is set up during f.BeforeEach, so len(ns) won't be 0 even if the test is skipped.

File: test/e2e/apimachinery/webhook.go (hunk 7 above)
Comment (janetkuo) [reply test/e2e/apimachinery/webhook.go]: > we need to call these cleanups after each `It` block

Why not just use `AfterEach`?

> @cheftako has some code in his PR that's doing the right thing I think.

Would you provide the link to the PR?

File: test/e2e/apimachinery/webhook.go (hunk 6 above)
Comment (janetkuo) [reply test/e2e/apimachinery/webhook.go]: Hm, probably a rebase mistake

File: test/e2e/apimachinery/webhook.go (hunk 8 above)
Comment (janetkuo) [reply test/e2e/apimachinery/webhook.go]:
```go
type Framework struct {
	Namespace                *v1.Namespace   // Every test has at least one namespace unless creation is skipped
}
```

File: test/e2e/apimachinery/webhook.go (hunk 7 above)
Comment (caesarxuchao) [reply test/e2e/apimachinery/webhook.go]: > Why not just use AfterEach?

That is what I meant :)

https://github.com/kubernetes/kubernetes/pull/55829

Comment (caesarxuchao) [reply test/e2e/apimachinery/webhook.go]: Thanks for fixing the CleanupAction! I'll go fix the webhook specific tests, that test shouldn't use CleanupAction from the beginning.

Comment (janetkuo) [reply test/e2e/apimachinery/webhook.go]: This PR is not trying to modify any e2e test, but just make AddCleanupAction work as intended (instead of noop).

File: test/e2e/apimachinery/webhook.go (hunk 6 above)
Comment (miaoyq) [reply test/e2e/apimachinery/webhook.go]: Thanks for review, sorry for this, it is indeed a rebase mistake. :(
I will correct it in another pr.

Comment (miaoyq) [reply test/e2e/apimachinery/webhook.go]: Seem @cheftako have fixed this, Thanks @cheftako :)
